### PR TITLE
UI Grab Bag of Fixes

### PR DESCRIPTION
**Pull request by MightyAx**
Visible Changes:
- DataAccess.Type change from select many to select one dropdown
- View mode support for dropdown lists
- View mode support for ontology lists
- Put back support for links in help text
- Question names and helper text for technologies [#327](https://github.com/ebi-ait/hca-ebi-dev-team/issues/327)
- Hide empty fields when using view mode
- Hide empty tabs when using view mode
- Hide tab bar if only one tab has non empty fields
- Login / Logout Footer Fix
- Support for multiple root ontologies in picker for [#327](https://github.com/ebi-ait/hca-ebi-dev-team/issues/327)
- Format Dates in view mode as: `Friday, January 1, 2021`
- Fix Broken Back button on project form
- Line break before wrangler email on homepage
- Fixed issue where chosen tab not respected when going from project view mode to edit mode

Invisible Changes (for devs):
- Default text fields with enum options to dropdown.
- Allow config overloads to be set by metadatakey rather than just key
- Fix probable bug custom object controls
- Add support for the removeEmptyFields param by using the unused isReadOnly setting.
- Refactored the global page, header and footer.
- Refactored visibleTabs based on review
- Linted files changes by this merge

Retracted Changes:
- ~~Change from **Do** the data require controlled access? to **Does** the data require controlled access?~~

Deployed to [Dev](https://dev.contribute.data.humancellatlas.org) for review
Requires: [ingest-core#38](https://github.com/ebi-ait/ingest-core/pull/38)

### CODE REVIEW

**Comment by MightyAx** [comment]
Known issues:
- [ ] the login/logout text in the footer of the page shows logout before you've finished logging in
- [ ] changing tab in view mode and then clicking edit does not necessarily send you to the right tab in edit mode
  - We could choose not to fix this since we need to repo rip out the tabs implantation for [#294](https://github.com/ebi-ait/hca-ebi-dev-team/issues/294)

**Comment by clairerye** [comment]
Theses changes are looking good to me.
Yah for not allowing 'all managed access' and 'all open access' at the same time!
I like the view mode having just the completed fields showing.
Non urgent, non important preference (I am not sure we need to specify midnight British summer time)
image

I think also if I am being pedantic on grammar "Do the data require controlled access?" is correct as data is plural.

File: client/src/app/metadata-schema-form/metadata-form/metadata-form.component.html [hunk 1]
```diff
@@ -1,10 +1,10 @@
 <form *ngIf="done" class="vf-form" [formGroup]="metadataForm.formGroup" (submit)="onSubmit($event)">
 
-  <mat-tab-group *ngIf="config.layout.tabs.length > 1"
+  <mat-tab-group *ngIf="visibleTabs(config.layout.tabs) > 1"
```
Comment (aaclan-ebi) [inline client/src/app/metadata-schema-form/metadata-form/metadata-form.component.html:3]: You could have 1 function to return a list of tabs to be displayed (visible) and remove the `*ngIf=showTab`.  I believe that would also improve the rendering of the template.

File: client/src/app/metadata-schema-form/metadata-form/metadata-form.component.ts [hunk 2]
```diff
@@ -54,6 +55,37 @@ export class MetadataFormComponent implements OnInit {
     this.done = true;
   }
 
+  showTab(tab: MetadataFormTab): boolean {
+    if (this.config.viewMode && this.config.removeEmptyFields && tab.items.length == 1) {
+      if (tab.key == tab.items[0]) {
+        let data = this.data;
+        let chain = true;
+
+        let keys = tab.key.replace('project.', '').split('.')
```
Comment (aaclan-ebi) [inline client/src/app/metadata-schema-form/metadata-form/metadata-form.component.ts:64]: `project` is hardcoded (i've made this same mistake before lol)

You could split and just shift the first element.
e.g.
```
 const keys = tab.key.split('.');
 keys.shift();

for (const key of keys)
```

Comment (aaclan-ebi) [reply client/src/app/metadata-schema-form/metadata-form/metadata-form.component.ts:64]: Not sure if you're familiar with command+option+l (https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html#reformat_code) to  auto reformat code (it's very handy!) It seems that this is showing some `==` vs `===` and `let` vs `const` usage issues.

File: client/src/app/metadata-schema-form/metadata-field-types/ontology-list-input/ontology-list-input.component.ts [hunk 3]
```diff
@@ -26,7 +26,7 @@ export class OntologyListInputComponent extends OntologyBaseComponent implements
   ngOnInit(): void {
     super.ngOnInit();
     const value = this.metadataFormService.cleanFormData(this.control.value);
-    this.value = this.metadataFormService.isEmpty(value) ? value : undefined;
+    this.value = this.metadataFormService.isEmpty(value) ? undefined : value;
```
Comment (aaclan-ebi) [inline client/src/app/metadata-schema-form/metadata-field-types/ontology-list-input/ontology-list-input.component.ts:29]: good catch

**Review by aaclan-ebi — COMMENTED** [review]
Hi Alexie, Thanks for making the changes. I have some suggestions! Please check.

I also noticed when you run the tests (`npm run test`) some tests are failing, i think those are unrelated to your changes but if you get a chance to fix them, that would be great! Thanks!

**Comment by MightyAx** [comment]
New Fixes:
- Login / Logout Footer Fix
- Support for multiple ontologies for [#327](https://github.com/ebi-ait/hca-ebi-dev-team/issues/327)
- Refactored the global page, header and footer.

File: client/src/app/metadata-schema-form/metadata-form/metadata-form.component.ts (hunk 2 above)
Comment (MightyAx) [reply client/src/app/metadata-schema-form/metadata-form/metadata-form.component.ts:64]: I have so many squiggles in the UI that I can't usually use that but will have a look.

**Comment by MightyAx** [comment]
Deployed latest to dev with the following ToDos before it can merge:
- [ ] Date / Time format to just show date
- [ ] Change back to **Do the data require controlled access**?
  - I hate the English language
- [ ] Make Alegria's Suggested changes
- [ ] Fix unit tests
- [ ] Current Issue changing tab in view mode and then clicking edit does not necessarily send you to the right tab in edit mode, **We could choose not to fix this since we need to repo rip out the tabs implementation for #294**

**Comment by MightyAx** [comment]
@clairerye 👍 👎 ?
Screenshot 2020-09-24 at 10 26 28
> **Date for release of the data for this project**
> Friday, January 1, 2021
> **Accession Time Frame**
> Tuesday, September 1, 2020

**Comment by mshadbolt** [comment]
I may be being pedantic but since this is a general grab bag, it has always messed with my ocd that the wrangler-team email on the front page gets wrapped on to a new line after `wrangler-` when the window is full width ie
```
This is a beta service, and we are constantly improving it.
Get in touch with our Data Wranglers with any questions about data submission: wrangler-
team@data.humancellatlas.org.
```
Would it be easy to make sure the full email address is always kept on the same line and not broken? If not an easy quick fix feel free to ignore and I can add the feedback for a subsequent round of fixes.

**Comment by MightyAx** [comment]
@mshadbolt
> The wrangler-team email on the front page gets wrapped on to a new line after `wrangler-` when the window is full width.
> ```
> This is a beta service, and we are constantly improving it.
> Get in touch with our Data Wranglers with any questions about data submission: wrangler-
> team@data.humancellatlas.org.
> ```
The width of the page is locked by the visual framework width, but we could:
- start a new line before the email address
- or make the mailto: link part of some of the other text like [Data Wranglers](mailto:wrangler-team@data.humancellatlas.org)

**Comment by mshadbolt** [comment]
I think I like the first option with the new line, just because sometimes a mailto doesn't work if people don't have a mail app set up so it could be nice to be able to see and copy/paste the address if needed.

Though I do realise with this option, if someone makes the window narrow enough it will wrap it again, but I think I'm ok with it

**Comment by MightyAx** [comment]
I've separated the lint and unit test issues to a new branch since it changes a lot of code and will be hard to review otherwise:
https://github.com/ebi-ait/ingest-ui/tree/cleanup/lint-and-tests

**Comment by MightyAx** [comment]
Last call for reviews on these changes!
deployed to [Dev](https://dev.contribute.data.humancellatlas.org)

**Comment by mshadbolt** [comment]
Is it possible to remove the ontology curie from the view in the drop down so that it is just a list of techniques without the (EFO:#####)?

**Review by aaclan-ebi — APPROVED** [review]
Looks good, Alexie!!! Thanks for applying the changes.

I quickly test the project view changes locally and noticed some (i'd say) non critical issues 

1. that _Contributor role ontology_ is being displayed even when empty.
2.  I also noticed that `/projects/349d2b8d-08b0-49c0-aa2d-d0ae56598f2d/project.content.project_core` doesn't change when clicking next / or selecting new tabs. I'm not actually sure how it's supposed to work.

**Comment by MightyAx** [comment]
> `/projects/349d2b8d-08b0-49c0-aa2d-d0ae56598f2d/project.content.project_core` doesn't change when clicking next / or selecting new tabs.

The project component (that hosts the view mode and lists submissions) changes the destination of the edit button to take you to the correct tab in edit mode, [here](https://github.com/ebi-ait/ingest-ui/pull/52/files#diff-e43e9130b6c13c8049962e97175614e0).

Once you land on the the edit page, the url param is used to set the right starting tab, [here, line 74](https://github.com/ebi-ait/ingest-ui/pull/52/files#diff-ffe2168853836df2c807be6b466a54cc).
**The url isn't updated to reflect as you navigate the tabs in edit mode though, is that something we can do? should we?**

> Contributor role ontology is being displayed even when empty.

I see, in fact all ontology list fields display this behaviour.

**Summary of outstanding issues and questions going forward (@clairerye for prioritisation):**
- [ ] Empty Ontology List fields aren't hidden in view mode.
  - _(Note: This could be another Pandora's box)_.
- [ ] Update ontology display to only show name and not EFO, [code here](https://github.com/ebi-ait/ingest-ui/blob/33b452016b07e9d30b0da11f39d640b0843f7897/client/src/app/metadata-schema-form/metadata-field-types/ontology-base/ontology-base.component.ts#L47)
  - This would affect all our ontology pickers when selecting from search, editing and in view mode.
  - Would we want the display to be based on whether you were a wrangler or not?

**Pull request merged**